File: libs/ai-kernel/src/models/prompt_template.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sqlx::{FromRow, types::Json};
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PromptVersion {
    pub version_number: i32,
    pub content: String,
    #[serde(default)]
    pub input_variables: Vec<String>,
    pub notes: String,
    pub created_at: DateTime<Utc>,
    pub created_by: Option<Uuid>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PromptTemplate {
    pub id: Uuid,
    pub name: String,
    pub description: String,
    pub category: String,
    pub status: String,
    pub tags: Vec<String>,
    pub latest_version_number: i32,
    pub current_version: PromptVersion,
    pub versions: Vec<PromptVersion>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ListPromptTemplatesResponse {
    pub data: Vec<PromptTemplate>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreatePromptTemplateRequest {
    pub name: String,
    #[serde(default)]
    pub description: String,
    #[serde(default = "default_prompt_category")]
    pub category: String,
    pub content: String,
    #[serde(default)]
    pub input_variables: Vec<String>,
    #[serde(default)]
    pub tags: Vec<String>,
    #[serde(default)]
    pub notes: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct UpdatePromptTemplateRequest {
    pub name: Option<String>,
    pub description: Option<String>,
    pub category: Option<String>,
    pub status: Option<String>,
    pub content: Option<String>,
    pub input_variables: Option<Vec<String>>,
    pub tags: Option<Vec<String>>,
    pub notes: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RenderPromptRequest {
    #[serde(default)]
    pub variables: Value,
    #[serde(default)]
    pub strict: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RenderPromptResponse {
    pub prompt_id: Uuid,
    pub version_number: i32,
    pub rendered_content: String,
    pub missing_variables: Vec<String>,
}

#[derive(Debug, Clone, FromRow)]
pub(crate) struct PromptTemplateRow {
    pub id: Uuid,
    pub name: String,
    pub description: String,
    pub category: String,
    pub status: String,
    pub tags: Json<Vec<String>>,
    pub latest_version_number: i32,
    pub versions: Json<Vec<PromptVersion>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl From<PromptTemplateRow> for PromptTemplate {
    fn from(value: PromptTemplateRow) -> Self {
        let current_version = value.versions.0.last().cloned().unwrap_or(PromptVersion {
            version_number: value.latest_version_number,
            content: String::new(),
            input_variables: Vec::new(),
            notes: String::new(),
            created_at: value.created_at,
            created_by: None,
        });

        Self {
            id: value.id,
            name: value.name,
            description: value.description,
            category: value.category,
            status: value.status,
            tags: value.tags.0,
            latest_version_number: value.latest_version_number,
            current_version,
            versions: value.versions.0,
            created_at: value.created_at,
            updated_at: value.updated_at,
        }
    }
}

fn default_prompt_category() -> String {
    "copilot".to_string()
}

File: openfoundry-go/libs/ai-kernel-go/models/prompt_template.go
package models

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// PromptVersion is one revision of a prompt template.
type PromptVersion struct {
	VersionNumber  int32      `json:"version_number"`
	Content        string     `json:"content"`
	InputVariables []string   `json:"input_variables,omitempty"`
	Notes          string     `json:"notes"`
	CreatedAt      time.Time  `json:"created_at"`
	CreatedBy      *uuid.UUID `json:"created_by"`
}

// PromptTemplate carries every version + the convenience pointer
// at the latest one (current_version).
type PromptTemplate struct {
	ID                   uuid.UUID       `json:"id"`
	Name                 string          `json:"name"`
	Description          string          `json:"description"`
	Category             string          `json:"category"`
	Status               string          `json:"status"`
	Tags                 []string        `json:"tags"`
	LatestVersionNumber  int32           `json:"latest_version_number"`
	CurrentVersion       PromptVersion   `json:"current_version"`
	Versions             []PromptVersion `json:"versions"`
	CreatedAt            time.Time       `json:"created_at"`
	UpdatedAt            time.Time       `json:"updated_at"`
}

type ListPromptTemplatesResponse struct {
	Data []PromptTemplate `json:"data"`
}

// CreatePromptTemplateRequest defaults: category="copilot".
type CreatePromptTemplateRequest struct {
	Name           string   `json:"name"`
	Description    *string  `json:"description,omitempty"`
	Category       *string  `json:"category,omitempty"`
	Content        string   `json:"content"`
	InputVariables []string `json:"input_variables,omitempty"`
	Tags           []string `json:"tags,omitempty"`
	Notes          *string  `json:"notes,omitempty"`
}

type UpdatePromptTemplateRequest struct {
	Name           *string   `json:"name,omitempty"`
	Description    *string   `json:"description,omitempty"`
	Category       *string   `json:"category,omitempty"`
	Status         *string   `json:"status,omitempty"`
	Content        *string   `json:"content,omitempty"`
	InputVariables *[]string `json:"input_variables,omitempty"`
	Tags           *[]string `json:"tags,omitempty"`
	Notes          *string   `json:"notes,omitempty"`
}

type RenderPromptRequest struct {
	Variables json.RawMessage `json:"variables,omitempty"`
	Strict    bool            `json:"strict"`
}

type RenderPromptResponse struct {
	PromptID         uuid.UUID `json:"prompt_id"`
	VersionNumber    int32     `json:"version_number"`
	RenderedContent  string    `json:"rendered_content"`
	MissingVariables []string  `json:"missing_variables"`
}

const DefaultPromptCategory = "copilot"
